### ISSUE TITLE
Some questions after reading the code

### ISSUE THREAD

**Issue by HuangBo-Terraloupe** [yhenon/keras-frcnn#153]
Hello everyone,
After I reading the code, I have few questions, hope can get answers. 
1, the first question is about the argument C.num_rois. There will be 300 region proposals sent into the Roi-pooling layer, is this C.num_rois define how many region proposals will be forwarded the classifier?
If yes, why? Because all proposals will after non-maximum suppression be decided, to be output or discarded.
 
2, the second question is about the input and output of RPN.
Line 110 in train.py: 
data_gen_train = data_generators.get_anchor_gt(train_imgs, classes_count, C, nn.get_img_output_length, K.image_dim_ordering(), mode='train')
Dose the generator calculate the GT of region proposals? Because for 	
X, Y, img_data = next(data_gen_train)
Y will be the label to train the RPN. What confuse me is, the shape of Y[0] here is (1, 38, 38, 18), and Y[1] is (1, 38, 38, 72).
But when you do 
P_rpn = model_rpn.predict_on_batch(X)
The shape of P_rpn[0] is (1, 38, 38, 9) and P_rpn[1] is (1, 38, 38, 36).
For my understanding, the num_kernel_dim should be 9. why Y[0] is 18 ?

3, One general question, why we need here sign positive or negative to every region proposal for training? 
			neg_samples = np.where(Y1[0, :, -1] == 1)
			pos_samples = np.where(Y1[0, :, -1] == 0)
Usually we just take care the label, then train it.

4, 
Line 189: X2, Y1, Y2, IouS = roi_helpers.calc_iou(R, img_data, C, class_mapping)
Line 232: loss_class = model_classifier.train_on_batch([X, X2[:, sel_samples, :]], [Y1[:, sel_samples, :], Y2[:, sel_samples, :]])

What are the X2, Y1, Y2 here? Want to make sure I understand.

**Comment by HuangBo-Terraloupe** [comment]
X is the input image, X2[:, sel_samples, :] should be the 32 chosen proposals, Y1[:, sel_samples, :] is the classification labels,  Y2[:, sel_samples, :] should be the regression GT, but the shape is (1, 32, 32), I do not understand.

**Comment by olaff09** [comment]
Hi @HuangBo-Terraloupe,

I have some of these questions myself, I will add comments on what I know:

1- Yes, 300 (num_rois) is the number of proposals sent to the classifier after doing non-max suppression, this is mostly for performance reason, you can change it as you need (You'll see how slower it is by just doubling that quantity). Those regions will be the "top 300" with highest probabilities of being an object.

2- I'd like to know why dimensions are different, too.

3- Positive and negative labeling is needed because region positions are not exact, we need a way for the  network to separate "good" regions from "almost good, but not enough" regions. Let me explain:
Positive = Regions that overlap a GT object by more than a threshold (0.7 lets say), meaning that the region contains a big part (or all) of the object, this is what we want.
Negative = Regions that overlap a GT object by less than a the threshold, meaning that they are not over an object or have only part of an object, so we don't want the network to classify these as positive.
This makes the network more resilient to object positions.

**Comment by yhenon** [comment]
Regarding (2), the array has twice the number of channels because we need to indicate whether a region should be used for backpropagating the loss. So we append that information in extra channels. It's kind of a dirty hack but does the trick.

**Comment by thealchemist89** [comment]
Hi @HuangBo-Terraloupe ,

I am not sure if I am right, but here is what I think:

4. X2 -> predicted ROI (x,y,w,h), Y1 -> Ground Truth class label, Y2 -> Regressed Bbox Coordinates and class labels.

Any verifications/corrections will be great for sharing! :)

**Comment by HuangBo-Terraloupe** [comment]
Thanks for the explanation!
And I have some further questions about the positive and negative samples.
In the paper, it seems we need these positive and negative sample because of the training of RPN.
And in the code here, it seems we use these positive and negative samples to get a batch of data for the training of classifier:
loss_class = model_classifier.train_on_batch([X, X2[:, sel_samples, :]], [Y1[:, sel_samples, :], Y2[:, sel_samples, :]]
And the batch size is the C.num_rois, half is positive samples, half is negative samples.
Why we have that difference?

**Comment by olaff09** [comment]
If you ask why are them half positive / half negative, I guess this is just a concern of balance,  there are papers where they use other proportions (e.g 2/3 positive, 1/3 neg) but this remains part of the field of study, I guess the "safe" bet is showing the same quantity for both pos and neg.

**Comment by dmitrysarov** [comment]
> Regarding (2), the array has twice the number of channels because we need to indicate whether a region should be used for backpropagating the loss. So we append that information in extra channels. It's kind of a dirty hack but does the trick.
Please, can you explain more about this trick or give some direction where to read?

**Comment by Tbarowski** [comment]
@yhenon: What causes (2) in the code? I think i understand the concept with the loss function but in the code all i see is a Dense Layer that is defined with outputs = classcount-1. What triggers keras to expect twice as many inputs in training as giving in testing?
